File: p938e/range_sum_bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
const Tree = require('binary_tree');
/**
 * @param {TreeNode} root
 * @param {number} L
 * @param {number} R
 * @return {number}
 */
var rangeSumBST = function (root, low, high) {
  let sum = 0;
  // Helper with closure of L, R, and sum
  const recHelper = (node) => {
    if (node == null) return;
    if (node.val >= low && node.val <= high) sum += node.val;
    if (node.val > low) recHelper(node.left);
    if (node.val < high) recHelper(node.right);
  };
  recHelper(root);
  return sum;
};
// TESTS
[
  {
    tree: '10,5,3,#,#,7,#,#,15,#,18,#,#',
    low: 7,
    high: 15,
    expected: 32,
  },
  {
    tree: '10,5,3,1,#,#,#,7,6,#,#,#,15,13,#,#,18,#,#',
    low: 6,
    high: 10,
    expected: 23,
  },
].forEach(({ tree, low, high, expected }) => {
  const actual = rangeSumBST(Tree.deserialize(tree), low, high);
  console.log('Range sum of', tree, low, high, '->', actual);
  console.assert(actual === expected);
});

File: p938e/range_sum_bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from local_packages.binary_tree import TreeNode


class Solution:
    def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
        if not root:
            return 0
        if root.val < low:
            return self.rangeSumBST(root.right, low, high)
        if root.val > high:
            return self.rangeSumBST(root.left, low, high)
        return (
            root.val
            + self.rangeSumBST(root.left, low, high)
            + self.rangeSumBST(root.right, low, high)
        )


# TESTS
for tree, low, high, expected in [
    ("10,5,3,#,#,7,#,#,15,#,18,#,#", 7, 15, 32),
    ("10,5,3,1,#,#,#,7,6,#,#,#,15,13,#,#,18,#,#", 6, 10, 23),
]:
    sol = Solution()
    actual = sol.rangeSumBST(TreeNode.deserialize(tree), low, high)
    print("Range sum of", tree, "->", actual)
    assert actual == expected
